src/types/sequence.rs

```diff
@@ -265,6 +208,312 @@ impl PySequence {
+    ///
+    /// This is equivalent to the Python statement `del self[i1:i2]`.
+    fn del_slice(&self, i1: usize, i2: usize) -> PyResult<()>;
+
+    /// Returns the number of occurrences of `value` in self, that is, return the
+    /// number of keys for which `self[key] == value`.
+    #[cfg(not(PyPy))]
+    fn count<V>(&self, value: V) -> PyResult<usize>
+    where
+        V: ToPyObject;
+
+    /// Determines if self contains `value`.
+    ///
+    /// This is equivalent to the Python expression `value in self`.
+    fn contains<V>(&self, value: V) -> PyResult<bool>
+    where
+        V: ToPyObject;
+
+    /// Returns the first index `i` for which `self[i] == value`.
+    ///
+    /// This is equivalent to the Python expression `self.index(value)`.
+    fn index<V>(&self, value: V) -> PyResult<usize>
+    where
+        V: ToPyObject;
+
+    /// Returns a fresh list based on the Sequence.
+    fn to_list(&self) -> PyResult<Py2<'py, PyList>>;
+
+    /// Returns a fresh tuple based on the Sequence.
+    fn to_tuple(&self) -> PyResult<Py2<'py, PyTuple>>;
+}
+
+impl<'py> PySequenceMethods<'py> for Py2<'py, PySequence> {
+    #[inline]
+    fn len(&self) -> PyResult<usize> {
+        let v = unsafe { ffi::PySequence_Size(self.as_ptr()) };
+        crate::err::error_on_minusone(self.py(), v)?;
+        Ok(v as usize)
+    }
+
+    #[inline]
+    fn is_empty(&self) -> PyResult<bool> {
+        self.len().map(|l| l == 0)
+    }
+
+    #[inline]
+    fn concat(&self, other: &Py2<'_, PySequence>) -> PyResult<Py2<'py, PySequence>> {
+        unsafe {
+            Py2::from_owned_ptr_or_err(
+                self.py(),
+                ffi::PySequence_Concat(self.as_ptr(), other.as_ptr()),
+            )
+            .map(|py2| py2.downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn repeat(&self, count: usize) -> PyResult<Py2<'py, PySequence>> {
+        unsafe {
+            Py2::from_owned_ptr_or_err(
+                self.py(),
+                ffi::PySequence_Repeat(self.as_ptr(), get_ssize_index(count)),
+            )
+            .map(|py2| py2.downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn in_place_concat(&self, other: &Py2<'_, PySequence>) -> PyResult<Py2<'py, PySequence>> {
+        unsafe {
+            Py2::from_owned_ptr_or_err(
+                self.py(),
+                ffi::PySequence_InPlaceConcat(self.as_ptr(), other.as_ptr()),
+            )
+            .map(|py2| py2.downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn in_place_repeat(&self, count: usize) -> PyResult<Py2<'py, PySequence>> {
+        unsafe {
+            Py2::from_owned_ptr_or_err(
+                self.py(),
+                ffi::PySequence_InPlaceRepeat(self.as_ptr(), get_ssize_index(count)),
+            )
+            .map(|py2| py2.downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn get_item(&self, index: usize) -> PyResult<Py2<'py, PyAny>> {
+        unsafe {
```

Comment:
Would it make sense to abstract out the repeating

```rust
Py2::from_owned_ptr_or_err(
    self.py(),
    f(self.as_ptr()),
)
.map(|py2| py2.downcast_into_unchecked())
```

using a for example a closure-based API with `f: FnOnce(*mut PyObject) -> R`?

Reply:
Yes, I think there could definitely be a `pub(crate)` helper to do this. I'll have a play 👍

Reply:
I've pushed a commit which adds `Py2::do_ffi_call` which has a generic return type with a bit of trait magic to map raw ffi return values into various safe wrappers. What do you think?